sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -72,6 +72,26 @@ case class CurrentTimestamp() extends LeafExpression with CodegenFallback {
 }
 
 /**
+ * Expression representing the current batch time, which is used by StreamExecution to
+ * 1. prevent optimizer from pushing this expression below a stateful operator
+ * 2. allow IncrementalExecution to substitute this expression with a Literal(timestamp)
+ *
+ * There is no code generation since this expression should be replaced with a literal.
+ */
+case class CurrentBatchTimestamp(timestamp: SQLTimestamp) extends LeafExpression
```

Comment:
Feel free to push back on this...
Since we're going to support both `CurrentTimestamp` and `CurrentDate`, I would recommend the following signature:
```scala
case class CurrentBatchTimestamp(timestampMs: Long, dataType: DataType) {

  def toLiteral: Literal = dataType match {
    case _: TimestampType => Literal(timestampMs * 1000, TimestampType)
    case _: DateType => Literal(DateTimeUtils.millisToDays(timestampMs), DateType)
  }
```